File: project/app/src/main/java/com/iven/musicplayergo/adapters/ArtistsAdapter.java
package com.iven.musicplayergo.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.text.Spanned;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.iven.musicplayergo.R;
import com.iven.musicplayergo.Utils;
import com.iven.musicplayergo.models.Artist;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class ArtistsAdapter extends RecyclerView.Adapter<ArtistsAdapter.SimpleViewHolder> {

    private final Context mContext;
    private final ArtistSelectedListener mArtistSelectedListener;
    private final HashMap<Integer, String> mIndexedPositions = new LinkedHashMap<>();
    private List<Artist> mArtists;

    public ArtistsAdapter(@NonNull final Context context, @NonNull final List<Artist> artists) {
        mContext = context;
        mArtists = artists;
        mArtistSelectedListener = (ArtistSelectedListener) mContext;
        generateIndexes();
    }

    public void setArtists(List<Artist> artists) {
        mArtists = artists;
        notifyDataSetChanged();
    }

    public String[] getIndexes() {
        int size = mIndexedPositions.values().size();
        return mIndexedPositions.values().toArray(new String[size]);
    }

    public int getIndexPosition(final int currentSection) {
        final int size = mIndexedPositions.keySet().size();
        return mIndexedPositions.keySet().toArray(new Integer[size])[currentSection];
    }

    private void generateIndexes() {
        try {
            for (int i = 0, size = mArtists.size(); i < size; i++) {
                String section = mArtists.get(i).getName().substring(0, 1).toUpperCase(Locale.getDefault());
                if (!mIndexedPositions.containsValue(section)) {
                    mIndexedPositions.put(i, section);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    @NonNull
    public SimpleViewHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {

        final View itemView = LayoutInflater.from(mContext)
                .inflate(R.layout.artist_item, parent, false);

        return new SimpleViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull final SimpleViewHolder holder, final int position) {

        Artist artist = mArtists.get(holder.getAdapterPosition());
        holder.title.setText(artist.getName());

        final Spanned spanned = Utils.buildSpanned(mContext.getString(R.string.artist_info, artist.getAlbums().size(), artist.getSongCount()));

        holder.albumCount.setText(spanned);
    }

    @Override
    public int getItemCount() {
        return mArtists.size();
    }

    public interface ArtistSelectedListener {
        void onArtistSelected(@NonNull final String artist, final boolean showPlayedArtist);
    }

    class SimpleViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener, View.OnTouchListener {

        final TextView title, albumCount;
        private boolean sArtistLongPressed = false;

        SimpleViewHolder(@NonNull final View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.artist);
            albumCount = itemView.findViewById(R.id.album_count);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(@NonNull final View v) {
            mArtistSelectedListener.onArtistSelected(mArtists.get(getAdapterPosition()).getName(), false);
        }

        @Override
        public boolean onLongClick(@NonNull final View v) {
            if (!sArtistLongPressed) {
                title.setSelected(true);
                sArtistLongPressed = true;
            }
            return true;
        }

        @Override
        @SuppressLint("ClickableViewAccessibility")
        public boolean onTouch(@NonNull final View v, @NonNull final MotionEvent event) {
            if (sArtistLongPressed && event.getAction() == MotionEvent.ACTION_UP || event.getAction() == MotionEvent.ACTION_OUTSIDE || event.getAction() == MotionEvent.ACTION_MOVE) {
                title.setSelected(false);
                sArtistLongPressed = false;
            }
            return false;
        }
    }
}

File: project/app/src/main/java/com/iven/musicplayergo/adapters/ArtistsAdapter.kt
package com.iven.musicplayergo.adapters

import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.iven.musicplayergo.R
import com.iven.musicplayergo.music.Music
import com.iven.musicplayergo.music.MusicUtils
import kotlinx.android.synthetic.main.artist_item.view.*
import java.util.*

class ArtistsAdapter(
    private val resources: Resources,
    artists: MutableList<String>,
    private val music: Map<String, Map<String?, List<Music>>>
) :
    RecyclerView.Adapter<ArtistsAdapter.ArtistsHolder>() {

    var onArtistClick: ((String?) -> Unit)? = null
    private var mArtists: List<String> = artists

    init {
        Collections.sort(mArtists, String.CASE_INSENSITIVE_ORDER)
    }

    fun setArtists(artists: List<String>) {
        mArtists = artists
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ArtistsHolder {
        return ArtistsHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.artist_item,
                parent,
                false
            )
        )
    }

    override fun getItemCount(): Int {
        return mArtists.size
    }

    override fun onBindViewHolder(holder: ArtistsHolder, position: Int) {
        val artist = mArtists[holder.adapterPosition]
        val albums = music.getValue(artist)
        holder.bindItems(
            mArtists[holder.adapterPosition],
            albums.keys.size,
            MusicUtils.getArtistSongsCount(albums)
        )
    }

    inner class ArtistsHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bindItems(title: String, albumCount: Int, songCount: Int) {
            itemView.artist.text = title
            itemView.album_count.text =
                MusicUtils.buildSpanned(
                    resources.getString(
                        R.string.artist_info,
                        albumCount,
                        songCount
                    )
                )
            itemView.setOnClickListener { onArtistClick?.invoke(mArtists[adapterPosition]) }
        }
    }
}
